### ISSUE TITLE
Gradient checkpointing fails to backprop in some cases

### ISSUE THREAD

**Issue by stephenroller** [pytorch/pytorch#33962]
## 🐛 Bug

I'm attempting to use `torch.utils.checkpoint.checkpoint`. I've found that it fails to properly call of `CheckpointFunction.backward` in some cases.

## To Reproduce

```python
import torch
import torch.nn as nn
from torch.utils.checkpoint import checkpoint


class Model(nn.Module):
    def __init__(self, n: int, use_cp: bool, allow_first: bool, allow_last: bool):
        super().__init__()
        self.layers = nn.ModuleList()
        self.n = n
        self.use_cp = use_cp
        self.allow_first = allow_first
        self.allow_last = allow_last
        for i in range(self.n):
            self.layers.append(nn.Linear(256, 256))

    def forward(self, x):
        for i in range(self.n):
            if (
                not self.use_cp
                or (i == 0 and not self.allow_first)
                or (i == self.n - 1 and not self.allow_last)
            ):
                print("No checkpoint", i)
                x = self.layers[i](x)
            else:
                print("Checkpointing", i)
                x = checkpoint(self.layers[i], x)
        return x


def test(use_cp, first, last):
    model = Model(4, use_cp, first, last).cuda()
    x = torch.randn(17, 256).cuda()
    loss = model(x).sum()
    try:
        loss.backward()
    except RuntimeError:
        return "RuntimeError"
    return sum([p.grad is None for p in model.parameters()])


print("None grads with NO grad checkpoint:", test(False, False, False))
print()
print("None grads with ALL grad checkpoint (1..n):", test(True, True, True))
print()
print("None grads with grad checkpoint (no first; 2..n):", test(True, False, True))
print()
print("None grads with grad checkpoint (no last; 1..n-1):", test(True, True, False))
print()
print("None grads with grad checkpoint (neither; 2..n-1):", test(True, False, False))
print()
```

Produces the output:
```
No checkpoint 0
No checkpoint 1
No checkpoint 2
No checkpoint 3
None grads with NO grad checkpoint: 0

Checkpointing 0
/private/home/roller/.conda/envs/chat/lib/python3.7/site-packages/torch/utils/checkpoint.py:25: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn("None of the inputs have requires_grad=True. Gradients will be None")
Checkpointing 1
Checkpointing 2
Checkpointing 3
None grads with ALL grad checkpoint (1..n): RuntimeError

No checkpoint 0
Checkpointing 1
Checkpointing 2
Checkpointing 3
None grads with grad checkpoint (no first; 2..n): 0

Checkpointing 0
Checkpointing 1
Checkpointing 2
No checkpoint 3
None grads with grad checkpoint (no last; 1..n-1): 6

No checkpoint 0
Checkpointing 1
Checkpointing 2
No checkpoint 3
None grads with grad checkpoint (neither; 2..n-1): 0

```

## Expected behavior

All test cases should produce 0 None grads, or at least RuntimeErrors. The fourth case, in which only 3/4 layers have gradients produces is VERY concerning.

## Environment

```
$ python -m torch.utils.collect_env
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration:
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 418.116.00
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] numpy-stubs==0.0.1
[pip] pytorch-pretrained-bert==0.6.2
[pip] pytorch-transformers==1.1.0
[pip] torch==1.4.0
[pip] torchtext==0.4.0
[pip] torchvision==0.5.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-service               2.0.2            py37h7b6447c_0
[conda] mkl_fft                   1.0.14           py37ha843d7b_0
[conda] mkl_random                1.0.2            py37hd81dba3_0
[conda] pytorch                   1.3.1           py3.7_cuda10.0.130_cudnn7.6.3_0    pytorch
[conda] pytorch-pretrained-bert   0.6.1                     <pip>
[conda] pytorch-pretrained-bert   0.6.2                     <pip>
[conda] pytorch-transformers      1.1.0                     <pip>
[conda] torch                     1.4.0                     <pip>
[conda] torchtext                 0.4.0                     <pip>
[conda] torchvision               0.4.2                py37_cu100    pytorch
[conda] torchvision               0.5.0                     <pip>
```

**Comment by stephenroller** [comment]
And with nightly:

```
$ python test.py
No checkpoint 0
No checkpoint 1
No checkpoint 2
No checkpoint 3
None grads with NO grad checkpoint: 0

Checkpointing 0
/private/home/roller/.conda/envs/torch_nightly_20200228/lib/python3.7/site-packages/torch/utils/checkpoint.py:25: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn("None of the inputs have requires_grad=True. Gradients will be None")
Checkpointing 1
Checkpointing 2
Checkpointing 3
None grads with ALL grad checkpoint (1..n): RuntimeError

No checkpoint 0
Checkpointing 1
Checkpointing 2
Checkpointing 3
None grads with grad checkpoint (no first; 2..n): 0

Checkpointing 0
Checkpointing 1
Checkpointing 2
No checkpoint 3
None grads with grad checkpoint (no last; 1..n-1): 6

No checkpoint 0
Checkpointing 1
Checkpointing 2
No checkpoint 3
None grads with grad checkpoint (neither; 2..n-1): 0

$ python -m torch.utils.collect_env
Collecting environment information...
PyTorch version: 1.5.0.dev20200228
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration:
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 418.116.00
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.5.0.dev20200228
[pip] torchvision==0.6.0a0+bd230a0
[conda] blas                      1.0                         mkl
[conda] mkl                       2020.0                      166
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.0.15           py37ha843d7b_0
[conda] mkl_random                1.1.0            py37hd6b4f25_0
[conda] pytorch                   1.5.0.dev20200228 py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch-nightly
[conda] torchvision               0.6.0.dev20200228      py37_cu101    pytorch-nightly
```

**Comment by stephenroller** [comment]
It's worth noting that i have a real case with a transformer, where the issue seems much more nuanced than "first/last layers". I have not yet identified the pattern.

**Comment by albanD** [comment]
This is expected behavior I'm afraid.

- The first one raise an error because, since you checkpoint all the way to the end, you try to backward something that does not require gradient and so it fails.
- The second one works because the first module is not checkpointed and it's output requires grad. So the next checkpoint will work
- The third one, the first modules won't work, but since the last one is not checkpointed, it produces an output that requires gradient and so the backward does not through a runtime error.
- The last one works for the same reason as the second one.

Making the input to your model require gradients will fix that no?

Also looks like we have a warn-once for the "Be careful gradients are going to be None". Maybe we should make it more verbose :D

**Comment by stephenroller** [comment]
Thanks again for the thorough answer. This is a failure of me boiling down my test case properly. I have operations before/after and my actual checkpoints.

I've found the error boils down to some LayerNorms, and possibly some view operations I'm doing in there. That makes me think this might be #26546. I will check back when I figure more out.

**Comment by stephenroller** [comment]
Alright well I think I've sorted out out, so I'll close it.

Something about doing this operation like this:

https://github.com/facebookresearch/ParlAI/blob/b927d8e4f2a4d18cbc797c78d38d7506e08e6426/parlai/agents/transformer/modules.py#L43-L50

inside this block:

https://github.com/facebookresearch/ParlAI/blob/b927d8e4f2a4d18cbc797c78d38d7506e08e6426/parlai/agents/transformer/modules.py#L595-L604

Failed. As in I checkpointed the call to the lower chunk of code and when i commented out the calls to `_normalize`, things were fine.

Switching the layernorm to not use `view` operations seemed to make the problem go away (though I'm not sure my network is correct anymore but yolo)
